File: packages/agent/mqtt.go
package main

// Minimal MQTT 3.1.1 client over TCP/TLS.
//
// Why hand-rolled instead of pulling in a library: this carries signaling only
// — a handful of small JSON messages per session. The subset needed is
// CONNECT / SUBSCRIBE / PUBLISH(QoS 0) / PINGREQ / DISCONNECT, which is a few
// hundred lines. It also keeps the agent dependency-free, matching the browser
// side (packages/component/src/mqtt.ts), and lets the two be compared line by
// line.
//
// Note the transport differs from the browser on purpose: a browser must use
// MQTT-over-WebSocket, while a native process can use plain MQTT over TCP or
// TLS. Both reach the same broker and the same topics, so they interoperate.
//
// Deliberately omitted: QoS 1/2, retained messages, wills, reconnection.
//
// Spec: MQTT 3.1.1 (OASIS) — §2.2 fixed header, §3.1 CONNECT, §3.3 PUBLISH,
// §3.8 SUBSCRIBE, §3.12 PINGREQ.

import (
	"bufio"
	"crypto/tls"
	"encoding/binary"
	"errors"
	"fmt"
	"io"
	"net"
	"net/url"
	"strings"
	"sync"
	"time"
)

const (
	pktConnect    byte = 1
	pktConnack    byte = 2
	pktPublish    byte = 3
	pktSubscribe  byte = 8
	pktSuback     byte = 9
	pktPingreq    byte = 12
	pktPingresp   byte = 13
	pktDisconnect byte = 14
)

type mqttMessage struct {
	topic   string
	payload []byte
}

type mqttClient struct {
	conn     net.Conn
	reader   *bufio.Reader
	mu       sync.Mutex
	packetID uint16

	incoming chan mqttMessage
	connack  chan error
	suback   chan uint16

	done      chan struct{}
	closeOnce sync.Once
	failMu    sync.Mutex
	failErr   error
}

// dialMQTT connects and completes the CONNECT/CONNACK handshake.
//
// Accepted schemes: mqtt:// (plain) and mqtts:// (TLS). Default ports 1883/8883.
func dialMQTT(rawURL, clientID string, keepalive time.Duration) (*mqttClient, error) {
	u, err := url.Parse(rawURL)
	if err != nil {
		return nil, fmt.Errorf("parse mqtt url: %w", err)
	}

	host := u.Host
	if u.Port() == "" {
		if u.Scheme == "mqtts" {
			host = net.JoinHostPort(u.Hostname(), "8883")
		} else {
			host = net.JoinHostPort(u.Hostname(), "1883")
		}
	}

	var conn net.Conn
	switch strings.ToLower(u.Scheme) {
	case "mqtt", "tcp":
		conn, err = net.DialTimeout("tcp", host, 10*time.Second)
	case "mqtts", "ssl", "tls":
		dialer := &net.Dialer{Timeout: 10 * time.Second}
		conn, err = tls.DialWithDialer(dialer, "tcp", host, &tls.Config{
			ServerName: u.Hostname(),
		})
	default:
		return nil, fmt.Errorf("unsupported mqtt scheme %q (use mqtt:// or mqtts://)", u.Scheme)
	}
	if err != nil {
		return nil, fmt.Errorf("dial mqtt %s: %w", host, err)
	}

	c := &mqttClient{
		conn:     conn,
		reader:   bufio.NewReader(conn),
		incoming: make(chan mqttMessage, 64),
		connack:  make(chan error, 1),
		suback:   make(chan uint16, 4),
		done:     make(chan struct{}),
	}

	go c.readLoop()
	go c.pingLoop(keepalive)

	if err := c.sendConnect(clientID, uint16(keepalive.Seconds())); err != nil {
		c.Close()
		return nil, err
	}
	select {
	case err := <-c.connack:
		if err != nil {
			c.Close()
			return nil, err
		}
	case <-time.After(15 * time.Second):
		c.Close()
		return nil, errors.New("mqtt: timed out waiting for CONNACK")
	case <-c.done:
		c.Close()
		return nil, c.Err()
	}

	return c, nil
}

func (c *mqttClient) Err() error {
	c.failMu.Lock()
	defer c.failMu.Unlock()
	if c.failErr == nil {
		return errors.New("mqtt: connection closed")
	}
	return c.failErr
}

func (c *mqttClient) Messages() <-chan mqttMessage { return c.incoming }

func (c *mqttClient) Done() <-chan struct{} { return c.done }

// Subscribe issues a SUBSCRIBE and waits for its SUBACK.
func (c *mqttClient) Subscribe(topic string) error {
	c.mu.Lock()
	c.packetID++
	id := c.packetID
	c.mu.Unlock()

	body := make([]byte, 0, 2+len(topic)+3)
	body = append(body, byte(id>>8), byte(id))
	body = append(body, encodeMQTTString(topic)...)
	body = append(body, 0x00) // requested QoS 0

	if err := c.write(buildPacket(pktSubscribe, 0x02, body)); err != nil {
		return err
	}

	select {
	case got := <-c.suback:
		if got != id {
			return fmt.Errorf("mqtt: SUBACK for unexpected packet id %d", got)
		}
		return nil
	case <-time.After(15 * time.Second):
		return errors.New("mqtt: timed out waiting for SUBACK")
	case <-c.done:
		return c.Err()
	}
}

// Publish sends a QoS 0 message: fire and forget.
func (c *mqttClient) Publish(topic string, payload []byte) error {
	body := make([]byte, 0, len(topic)+len(payload)+2)
	body = append(body, encodeMQTTString(topic)...)
	body = append(body, payload...)
	return c.write(buildPacket(pktPublish, 0x00, body))
}

func (c *mqttClient) Close() {
	c.closeOnce.Do(func() {
		_ = c.write(buildPacket(pktDisconnect, 0x00, nil))
		close(c.done)
		_ = c.conn.Close()
	})
}

// ------------------------------------------------------------------ internals

func (c *mqttClient) fail(err error) {
	c.failMu.Lock()
	if c.failErr == nil {
		c.failErr = err
	}
	c.failMu.Unlock()
	c.closeOnce.Do(func() {
		close(c.done)
		_ = c.conn.Close()
	})
}

func (c *mqttClient) write(packet []byte) error {
	c.mu.Lock()
	defer c.mu.Unlock()
	if err := c.conn.SetWriteDeadline(time.Now().Add(10 * time.Second)); err != nil {
		return err
	}
	_, err := c.conn.Write(packet)
	return err
}

func (c *mqttClient) sendConnect(clientID string, keepaliveSeconds uint16) error {
	body := make([]byte, 0, 32+len(clientID))
	body = append(body, encodeMQTTString("MQTT")...)
	body = append(body, 0x04)             // protocol level 4 = MQTT 3.1.1
	body = append(body, 0x02)             // connect flags: clean session
	body = append(body, byte(keepaliveSeconds>>8), byte(keepaliveSeconds))
	body = append(body, encodeMQTTString(clientID)...)
	return c.write(buildPacket(pktConnect, 0x00, body))
}

func (c *mqttClient) readLoop() {
	for {
		header, err := c.reader.ReadByte()
		if err != nil {
			c.fail(err)
			return
		}
		length, err := readVarint(c.reader)
		if err != nil {
			c.fail(err)
			return
		}
		body := make([]byte, length)
		if _, err := io.ReadFull(c.reader, body); err != nil {
			c.fail(err)
			return
		}

		switch header >> 4 {
		case pktConnack:
			if len(body) < 2 {
				c.connack <- errors.New("mqtt: malformed CONNACK")
				continue
			}
			if code := body[1]; code != 0 {
				c.connack <- fmt.Errorf("mqtt: CONNACK refused, code=%d", code)
				continue
			}
			c.connack <- nil

		case pktSuback:
			if len(body) < 2 {
				continue
			}
			select {
			case c.suback <- binary.BigEndian.Uint16(body[:2]):
			default:
			}

		case pktPublish:
			msg, err := parsePublish(header&0x0f, body)
			if err != nil {
				continue
			}
			select {
			case c.incoming <- msg:
			case <-c.done:
				return
			}

		case pktPingresp:
			// keepalive acknowledged; nothing to do
		}
	}
}

func (c *mqttClient) pingLoop(keepalive time.Duration) {
	if keepalive <= 0 {
		return
	}
	interval := keepalive / 2
	if interval < 5*time.Second {
		interval = 5 * time.Second
	}
	ticker := time.NewTicker(interval)
	defer ticker.Stop()
	for {
		select {
		case <-ticker.C:
			if err := c.write(buildPacket(pktPingreq, 0x00, nil)); err != nil {
				return
			}
		case <-c.done:
			return
		}
	}
}

func parsePublish(flags byte, body []byte) (mqttMessage, error) {
	if len(body) < 2 {
		return mqttMessage{}, errors.New("mqtt: truncated PUBLISH")
	}
	topicLen := int(binary.BigEndian.Uint16(body[:2]))
	offset := 2 + topicLen
	if len(body) < offset {
		return mqttMessage{}, errors.New("mqtt: truncated PUBLISH topic")
	}
	topic := string(body[2:offset])

	// QoS 0 has no packet identifier; QoS 1/2 carry two more bytes.
	if qos := (flags >> 1) & 0x03; qos > 0 {
		offset += 2
	}
	if offset > len(body) {
		offset = len(body)
	}
	return mqttMessage{topic: topic, payload: body[offset:]}, nil
}

// encodeMQTTString writes a UTF-8 string with a 2-byte big-endian length.
func encodeMQTTString(s string) []byte {
	out := make([]byte, 2+len(s))
	binary.BigEndian.PutUint16(out, uint16(len(s)))
	copy(out[2:], s)
	return out
}

// encodeVarint encodes the MQTT "remaining length": 7 bits per byte, high bit
// set while more bytes follow.
func encodeVarint(n int) []byte {
	var out []byte
	for {
		b := byte(n % 128)
		n /= 128
		if n > 0 {
			b |= 0x80
		}
		out = append(out, b)
		if n == 0 {
			return out
		}
	}
}

func readVarint(r *bufio.Reader) (int, error) {
	value := 0
	multiplier := 1
	for i := 0; ; i++ {
		if i > 3 {
			return 0, errors.New("mqtt: malformed remaining length")
		}
		b, err := r.ReadByte()
		if err != nil {
			return 0, err
		}
		value += int(b&0x7f) * multiplier
		if b&0x80 == 0 {
			return value, nil
		}
		multiplier *= 128
	}
}

func buildPacket(packetType, flags byte, body []byte) []byte {
	length := encodeVarint(len(body))
	out := make([]byte, 0, 1+len(length)+len(body))
	out = append(out, packetType<<4|flags)
	out = append(out, length...)
	out = append(out, body...)
	return out
}

File: packages/component/src/mqtt.ts
/**
 * Minimal MQTT 3.1.1 client over WebSocket.
 *
 * Why hand-rolled instead of `mqtt.js`: this is used for *signaling only*\u2014a
 * handful of small JSON messages per session. A full client brings a bundle, a
 * dependency, and an API surface we do not need. The subset implemented here is
 * CONNECT / SUBSCRIBE / PUBLISH(QoS 0) / PINGREQ / DISCONNECT, which is about
 * 200 lines and works identically in the browser and in Node (both expose a
 * global `WebSocket`).
 *
 * Deliberately omitted: QoS 1/2, retained messages, wills, TLS options beyond
 * the WebSocket URL, reconnection (the caller drives retries).
 *
 * Spec: MQTT 3.1.1 (OASIS), sections 2.2 (fixed header), 3.1, 3.3, 3.8, 3.12.
 */

const enum PacketType {
  CONNECT = 1,
  CONNACK = 2,
  PUBLISH = 3,
  PUBACK = 4,
  SUBSCRIBE = 8,
  SUBACK = 9,
  PINGREQ = 12,
  PINGRESP = 13,
  DISCONNECT = 14,
}

const textEncoder = new TextEncoder();
const textDecoder = new TextDecoder();

export interface MqttClientOptions {
  /** e.g. `wss://broker.emqx.io:8084/mqtt` */
  url: string;
  clientId?: string;
  /** Keepalive in seconds. Default 30. */
  keepalive?: number;
  username?: string;
  password?: string;
}

export type MqttMessageHandler = (topic: string, payload: Uint8Array) => void;

/** Encode a UTF-8 string with a 2-byte big-endian length prefix. */
function encodeString(value: string): Uint8Array {
  const bytes = textEncoder.encode(value);
  const out = new Uint8Array(2 + bytes.length);
  out[0] = (bytes.length >> 8) & 0xff;
  out[1] = bytes.length & 0xff;
  out.set(bytes, 2);
  return out;
}

/** MQTT "remaining length": 7 bits per byte, high bit = continuation. */
function encodeVarint(value: number): Uint8Array {
  const out: number[] = [];
  let remaining = value;
  do {
    let byte = remaining % 128;
    remaining = Math.floor(remaining / 128);
    if (remaining > 0) byte |= 0x80;
    out.push(byte);
  } while (remaining > 0);
  return new Uint8Array(out);
}

function decodeVarint(bytes: Uint8Array, offset: number): { value: number; bytes: number } | null {
  let multiplier = 1;
  let value = 0;
  let index = 0;
  for (;;) {
    if (offset + index >= bytes.length) return null;
    if (index > 3) throw new Error("malformed MQTT remaining length");
    const byte = bytes[offset + index]!;
    value += (byte & 0x7f) * multiplier;
    index += 1;
    if ((byte & 0x80) === 0) break;
    multiplier *= 128;
  }
  return { value, bytes: index };
}

function buildPacket(type: PacketType, flags: number, body: Uint8Array): Uint8Array {
  const length = encodeVarint(body.length);
  const out = new Uint8Array(1 + length.length + body.length);
  out[0] = (type << 4) | flags;
  out.set(length, 1);
  out.set(body, 1 + length.length);
  return out;
}

function concat(parts: Uint8Array[]): Uint8Array {
  const total = parts.reduce((sum, part) => sum + part.length, 0);
  const out = new Uint8Array(total);
  let offset = 0;
  for (const part of parts) {
    out.set(part, offset);
    offset += part.length;
  }
  return out;
}

export class MqttClient {
  private readonly options: Required<Pick<MqttClientOptions, "url" | "keepalive">> &
    Omit<MqttClientOptions, "url" | "keepalive">;
  private socket: WebSocket | null = null;
  private buffer: Uint8Array = new Uint8Array(0);
  private pingTimer: ReturnType<typeof setInterval> | null = null;
  private connectResolve: (() => void) | null = null;
  private connectReject: ((error: Error) => void) | null = null;
  private pendingSubscribes = new Map<number, () => void>();
  private nextPacketId = 1;
  private messageHandlers = new Set<MqttMessageHandler>();
  private closeHandlers = new Set<(error?: Error) => void>();

  constructor(options: MqttClientOptions) {
    this.options = { keepalive: 30, ...options };
  }

  get connected(): boolean {
    return this.socket?.readyState === WebSocket.OPEN;
  }

  /**
   * Resolves once the broker has accepted the connection (CONNACK).
   *
   * A timeout is mandatory, not a nicety: a blocked or blackholed WebSocket
   * never fires an error, so without one the promise stays pending forever and
   * the UI spins with nothing to report.
   */
  connect(timeoutMs = 15000): Promise<void> {
    if (this.socket) throw new Error("MqttClient.connect() called twice");

    return new Promise<void>((resolve, reject) => {
      const timer = setTimeout(() => {
        this.fail(
          new Error(
            `MQTT connect timed out after ${timeoutMs}ms — ${this.options.url} ` +
              `is unreachable or blocked`,
          ),
        );
      }, timeoutMs);
      const settle = <T>(fn: (value: T) => void) => (value: T) => {
        clearTimeout(timer);
        fn(value);
      };
      this.connectResolve = settle(resolve);
      this.connectReject = settle(reject);

      // MQTT-over-WebSocket requires the "mqtt" subprotocol (MQTT 3.1.1 spec
      // §6.0). Brokers reject the handshake outright if it is omitted, which
      // surfaces as a generic WebSocket error rather than a useful message.
      const socket = new WebSocket(this.options.url, "mqtt");
      socket.binaryType = "arraybuffer";
      this.socket = socket;

      socket.addEventListener("open", () => this.sendConnect());
      socket.addEventListener("message", (event) => {
        const data = event.data;
        if (typeof data === "string") return; // MQTT is binary-only here
        this.onBytes(new Uint8Array(data as ArrayBuffer));
      });
      socket.addEventListener("error", () => {
        this.fail(
          new Error(
            `MQTT socket error — ${this.options.url} is unreachable or blocked`,
          ),
        );
      });
      socket.addEventListener("close", () => {
        this.stopPing();
        const wasConnected = this.connectResolve === null;
        this.connectResolve = null;
        this.connectReject = null;
        const error = wasConnected
          ? undefined
          : new Error(`MQTT connection closed before CONNACK — ${this.options.url}`);
        for (const handler of this.closeHandlers) handler(error);
        if (!wasConnected) this.fail(error!);
      });
    });
  }

  onMessage(handler: MqttMessageHandler): void {
    this.messageHandlers.add(handler);
  }

  onClose(handler: (error?: Error) => void): void {
    this.closeHandlers.add(handler);
  }

  subscribe(topic: string): Promise<void> {
    const packetId = this.nextPacketId++;
    const body = concat([
      new Uint8Array([(packetId >> 8) & 0xff, packetId & 0xff]),
      encodeString(topic),
      new Uint8Array([0x00]), // requested QoS 0
    ]);
    return new Promise<void>((resolve) => {
      this.pendingSubscribes.set(packetId, resolve);
      this.send(buildPacket(PacketType.SUBSCRIBE, 0x02, body));
    });
  }

  /** QoS 0 publish: fire and forget. */
  publish(topic: string, payload: Uint8Array | string): void {
    const bytes = typeof payload === "string" ? textEncoder.encode(payload) : payload;
    this.send(buildPacket(PacketType.PUBLISH, 0x00, concat([encodeString(topic), bytes])));
  }

  close(): void {
    this.stopPing();
    if (this.socket?.readyState === WebSocket.OPEN) {
      try {
        this.socket.send(buildPacket(PacketType.DISCONNECT, 0x00, new Uint8Array(0)));
      } catch {
        /* ignore */
      }
    }
    this.socket?.close();
    this.socket = null;
  }

  // ---------------------------------------------------------------- internals

  private send(packet: Uint8Array): void {
    if (this.socket?.readyState !== WebSocket.OPEN) {
      throw new Error("MQTT socket is not open");
    }
    this.socket.send(packet);
  }

  private sendConnect(): void {
    const flags = 0x02; // clean session
    const parts: Uint8Array[] = [
      encodeString("MQTT"),
      new Uint8Array([0x04, flags]),
      new Uint8Array([(this.options.keepalive >> 8) & 0xff, this.options.keepalive & 0xff]),
      encodeString(this.options.clientId ?? `etw-${Math.random().toString(16).slice(2, 12)}`),
    ];
    if (this.options.username !== undefined) {
      parts.push(encodeString(this.options.username));
      parts.push(encodeString(this.options.password ?? ""));
    }
    this.send(buildPacket(PacketType.CONNECT, 0x00, concat(parts)));
  }

  private onBytes(chunk: Uint8Array): void {
    this.buffer = this.buffer.length === 0 ? chunk : concat([this.buffer, chunk]);

    for (;;) {
      if (this.buffer.length < 2) return;
      const decoded = decodeVarint(this.buffer, 1);
      if (decoded === null) return;
      const headerBytes = 1 + decoded.bytes;
      const total = headerBytes + decoded.value;
      if (this.buffer.length < total) return;

      const header = this.buffer[0]!;
      const body = this.buffer.subarray(headerBytes, total);
      this.buffer = this.buffer.subarray(total);
      this.handlePacket(header >> 4, header & 0x0f, body);
    }
  }

  private handlePacket(type: number, flags: number, body: Uint8Array): void {
    switch (type) {
      case PacketType.CONNACK: {
        const returnCode = body[1] ?? 0xff;
        if (returnCode !== 0) {
          this.fail(new Error(`MQTT CONNACK refused, code=${returnCode}`));
          return;
        }
        const resolve = this.connectResolve;
        this.connectResolve = null;
        this.connectReject = null;
        this.startPing();
        resolve?.();
        return;
      }

      case PacketType.SUBACK: {
        const packetId = ((body[0] ?? 0) << 8) | (body[1] ?? 0);
        const resolve = this.pendingSubscribes.get(packetId);
        this.pendingSubscribes.delete(packetId);
        resolve?.();
        return;
      }

      case PacketType.PUBLISH: {
        const topicLength = ((body[0] ?? 0) << 8) | (body[1] ?? 0);
        const topic = textDecoder.decode(body.subarray(2, 2 + topicLength));
        const qos = (flags >> 1) & 0x03;
        // QoS 0 has no packet identifier; QoS 1/2 skip two more bytes.
        const payloadOffset = 2 + topicLength + (qos > 0 ? 2 : 0);
        const payload = body.subarray(payloadOffset);
        for (const handler of this.messageHandlers) handler(topic, payload);
        return;
      }

      case PacketType.PINGRESP:
        return;

      default:
        // PUBACK/UNSUBACK etc. are not used by this client.
        return;
    }
  }

  private startPing(): void {
    this.stopPing();
    const intervalMs = Math.max(5, Math.floor(this.options.keepalive * 1000 * 0.5));
    this.pingTimer = setInterval(() => {
      if (this.socket?.readyState === WebSocket.OPEN) {
        try {
          this.socket.send(buildPacket(PacketType.PINGREQ, 0x00, new Uint8Array(0)));
        } catch {
          /* ignore */
        }
      }
    }, intervalMs);
  }

  private stopPing(): void {
    if (this.pingTimer !== null) {
      clearInterval(this.pingTimer);
      this.pingTimer = null;
    }
  }

  private fail(error: Error): void {
    const reject = this.connectReject;
    this.connectReject = null;
    this.connectResolve = null;
    reject?.(error);
  }
}
